internal/gadget-runner/gadgetrunner.go

```diff
@@ -0,0 +1,125 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package gadgetrunner handles initializing gadgets and installed operators before
+handing them over to a specified runtime.
+*/
+package gadgetrunner
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/parser"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/runtime"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+// GadgetRunner handles running gadgets by the gadget interface; it orchestrates the whole lifecycle of the gadget
+// instance and communicates with gadget and runtime.
+type GadgetRunner struct {
+	ctx          context.Context
+	id           string
+	gadget       gadgets.Gadget
+	gadgetParams *params.Params
+	runtime      runtime.Runtime
+	parser       parser.Parser
+	operators    operators.Operators
+	logger       logger.Logger
+	result       []byte
+	resultError  error
+}
+
+func NewGadgetRunner(
+	ctx context.Context,
+	id string,
+	runtime runtime.Runtime,
+	gadget gadgets.Gadget,
+	parser parser.Parser,
+	logger logger.Logger,
+) *GadgetRunner {
+	return &GadgetRunner{
+		ctx:     ctx,
+		id:      id,
+		gadget:  gadget,
+		runtime: runtime,
+		parser:  parser,
+		logger:  logger,
+	}
+}
+
+func (r *GadgetRunner) ID() string {
+	return r.id
+}
+
+func (r *GadgetRunner) Context() context.Context {
+	return r.ctx
+}
+
+func (r *GadgetRunner) Parser() parser.Parser {
+	return r.parser
+}
+
+func (r *GadgetRunner) Runtime() runtime.Runtime {
+	return r.runtime
+}
+
+func (r *GadgetRunner) Gadget() gadgets.Gadget {
+	return r.gadget
+}
+
+func (r *GadgetRunner) Operators() operators.Operators {
+	return r.operators
+}
+
+func (r *GadgetRunner) Logger() logger.Logger {
+	return r.logger
+}
+
+func (r *GadgetRunner) SetResult(result []byte, resultError error) {
+	r.result = result
+	r.resultError = resultError
+}
+
+func (r *GadgetRunner) GetResult() ([]byte, error) {
+	return r.result, r.resultError
+}
+
+func (r *GadgetRunner) GadgetParams() *params.Params {
+	return r.gadgetParams
+}
+
+// RunGadget is the main function of GadgetRunner and controls the lifecycle of the gadget
+func (r *GadgetRunner) RunGadget(
+	runtimeParams *params.Params,
+	operatorParamCollection params.Collection,
+	operatorPerGadgetParamCollection params.Collection,
+	gadgetParams *params.Params,
+) error {
+	r.gadgetParams = gadgetParams
+	r.operators = operators.GetOperatorsForGadget(r.gadget)
```

Comment:
Because that information can be gotten without any other knowledge of the environment, whereas all other parameters to `NewGadgetRunner()` can not. Does this make sense? It prevents having to deal with that beforehand.

Reply:
Hmm, we could move
```golang
r.operators = operators.GetOperatorsForGadget(r.gadget)
```
to `NewGadgetRunner()`, but I'd leave setting the gadgetParams for `RunGadget`, as that is where all settings are passed.

But now that I think about it, we could probably move everything to `NewGadgetRunner()`, because all `Params` are pointers anyways. I'll check if that works. Then `RunGadget()` wouldn't have any parameters anymore.

Reply:
At that point, do we still need `RunGadget()`?

Reply:
> At that point, do we still need `RunGadget()`?

We still need a reference to runner to get the result (and possibly an error) that might be stored in there. While that might somehow be manageable to rewrite, I'd also have to check whether that would have any effect on the kubectl and gadgettracermgr implementations (I haven't updated them for quite some time now). But if it's not too important, I would postpone more changes to this until we work on those parts.

Reply:
IMO, we can resolve this conversation.